The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16534401323/job/46766215185) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46766215185) [(plain)](https://github.com/rust-lang/rust/commit/01f0cafea63fcbb7e90a7f621cf7840ab6ecca99/checks/46766215185/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
95 
+ error: type `types::Priv` is more private than the item `types::ES`
+   --> $DIR/private-in-public-warn.rs:27:9
+    |
+ LL |         pub static ES: Priv;
+    |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
+ error: type `types::Priv` is more private than the item `types::ef1`
+   --> $DIR/private-in-public-warn.rs:28:9
+    |
+ LL |         pub fn ef1(arg: Priv);
+    |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
+ error: type `types::Priv` is more private than the item `types::ef2`
+   --> $DIR/private-in-public-warn.rs:29:9
+    |
+ LL |         pub fn ef2() -> Priv;
+    |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
---
- 
- error: type `types::Priv` is more private than the item `types::ES`
-   --> $DIR/private-in-public-warn.rs:27:9
-    |
- LL |         pub static ES: Priv;
-    |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
-    |
- note: but type `types::Priv` is only usable at visibility `pub(self)`
-   --> $DIR/private-in-public-warn.rs:9:5
-    |
- LL |     struct Priv;
-    |     ^^^^^^^^^^^
- 
- error: type `types::Priv` is more private than the item `types::ef1`
-   --> $DIR/private-in-public-warn.rs:28:9
-    |
- LL |         pub fn ef1(arg: Priv);
-    |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
-    |
- note: but type `types::Priv` is only usable at visibility `pub(self)`
-   --> $DIR/private-in-public-warn.rs:9:5
-    |
- LL |     struct Priv;
-    |     ^^^^^^^^^^^
- 
- error: type `types::Priv` is more private than the item `types::ef2`
-   --> $DIR/private-in-public-warn.rs:29:9
-    |
- LL |         pub fn ef2() -> Priv;
-    |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
-    |
- note: but type `types::Priv` is only usable at visibility `pub(self)`
-   --> $DIR/private-in-public-warn.rs:9:5
-    |
- LL |     struct Priv;
-    |     ^^^^^^^^^^^
397 
398 warning: bounds on generic parameters in type aliases are not enforced
399   --> $DIR/private-in-public-warn.rs:42:23

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:27:9
- LL |         pub static ES: Priv; //~ ERROR type `types::Priv` is more private than the item `types::ES`
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:28:9
- LL |         pub fn ef1(arg: Priv); //~ ERROR type `types::Priv` is more private than the item `types::ef1`
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:29:9
- LL |         pub fn ef2() -> Priv; //~ ERROR type `types::Priv` is more private than the item `types::ef2`
-   --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
+ error: type `types::Priv` is more private than the item `types::ES`
+   --> $DIR/private-in-public-warn.rs:27:9
+    |
+ LL |         pub static ES: Priv;
+    |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
+ error: type `types::Priv` is more private than the item `types::ef1`
+   --> $DIR/private-in-public-warn.rs:28:9
+    |
+ LL |         pub fn ef1(arg: Priv);
+    |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
+ error: type `types::Priv` is more private than the item `types::ef2`
+   --> $DIR/private-in-public-warn.rs:29:9
+    |
+ LL |         pub fn ef2() -> Priv;
+    |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
+    |
+ note: but type `types::Priv` is only usable at visibility `pub(self)`
+   --> $DIR/private-in-public-warn.rs:9:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+ 
---
To only update this specific test, also pass `--test-args privacy/private-in-public-warn.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/privacy/private-in-public-warn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/private-in-public-warn" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: type `types::Priv` is more private than the item `types::Alias`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:15:5
   |
LL |     pub type Alias = Priv; //~ ERROR type `types::Priv` is more private than the item `types::Alias`
   |     ^^^^^^^^^^^^^^ type alias `types::Alias` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^
note: the lint level is defined here
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:5:9
   |
LL | #![deny(private_interfaces, private_bounds)]
   |         ^^^^^^^^^^^^^^^^^^

error: type `types::Priv` is more private than the item `E::V1::0`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:17:12
   |
LL |         V1(Priv), //~ ERROR type `types::Priv` is more private than the item `E::V1::0`
   |            ^^^^ field `E::V1::0` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: type `types::Priv` is more private than the item `E::V2::field`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:18:14
   |
LL |         V2 { field: Priv }, //~ ERROR type `types::Priv` is more private than the item `E::V2::field`
   |              ^^^^^^^^^^^ field `E::V2::field` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: type `types::Priv` is more private than the item `Tr::C`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:21:9
   |
LL |         const C: Priv = Priv; //~ ERROR type `types::Priv` is more private than the item `Tr::C`
   |         ^^^^^^^^^^^^^ associated constant `Tr::C` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error[E0446]: private type `types::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:22:9
   |
LL |     struct Priv;
   |     ----------- `types::Priv` declared as private
...
LL |         type Alias = Priv; //~ ERROR private type `types::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error: type `types::Priv` is more private than the item `Tr::f1`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:23:9
   |
LL |         fn f1(arg: Priv) {} //~ ERROR type `types::Priv` is more private than the item `Tr::f1`
   |         ^^^^^^^^^^^^^^^^ associated function `Tr::f1` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: type `types::Priv` is more private than the item `Tr::f2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:24:9
   |
LL |         fn f2() -> Priv { panic!() } //~ ERROR type `types::Priv` is more private than the item `Tr::f2`
   |         ^^^^^^^^^^^^^^^ associated function `Tr::f2` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error[E0446]: private type `types::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:32:9
   |
LL |     struct Priv;
   |     ----------- `types::Priv` declared as private
...
LL |         type Alias = Priv; //~ ERROR private type `types::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error: type `types::Priv` is more private than the item `types::ES`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:27:9
   |
LL |         pub static ES: Priv; //~ ERROR type `types::Priv` is more private than the item `types::ES`
   |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: type `types::Priv` is more private than the item `types::ef1`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:28:9
   |
LL |         pub fn ef1(arg: Priv); //~ ERROR type `types::Priv` is more private than the item `types::ef1`
   |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: type `types::Priv` is more private than the item `types::ef2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:29:9
   |
LL |         pub fn ef2() -> Priv; //~ ERROR type `types::Priv` is more private than the item `types::ef2`
   |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
   |
note: but type `types::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:9:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error: trait `traits::PrivTr` is more private than the item `traits::Alias`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:42:5
   |
LL |     pub type Alias<T: PrivTr> = T; //~ ERROR trait `traits::PrivTr` is more private than the item `traits::Alias`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ type alias `traits::Alias` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^
note: the lint level is defined here
---

error: trait `traits::PrivTr` is more private than the item `traits::Tr1`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:44:5
   |
LL |     pub trait Tr1: PrivTr {} //~ ERROR trait `traits::PrivTr` is more private than the item `traits::Tr1`
   |     ^^^^^^^^^^^^^^^^^^^^^ trait `traits::Tr1` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits::PrivTr` is more private than the item `traits::Tr2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:45:5
   |
LL |     pub trait Tr2<T: PrivTr> {} //~ ERROR trait `traits::PrivTr` is more private than the item `traits::Tr2`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ trait `traits::Tr2` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits::PrivTr` is more private than the item `traits::Tr3::Alias`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:47:9
   |
LL |         type Alias: PrivTr;
   |         ^^^^^^^^^^^^^^^^^^ associated type `traits::Tr3::Alias` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits::PrivTr` is more private than the item `traits::Tr3::f`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:49:9
   |
LL |         fn f<T: PrivTr>(arg: T) {}
   |         ^^^^^^^^^^^^^^^^^^^^^^^ associated function `traits::Tr3::f` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits::PrivTr` is more private than the item `traits::Pub<T>`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:54:5
   |
LL |     impl<T: PrivTr> Pub<T> {} //~ ERROR trait `traits::PrivTr` is more private than the item `traits::Pub<T>`
   |     ^^^^^^^^^^^^^^^^^^^^^^ implementation `traits::Pub<T>` is reachable at visibility `pub(crate)`
   |
note: but trait `traits::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:37:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits_where::PrivTr` is more private than the item `traits_where::Alias`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:63:5
   |
LL |     pub type Alias<T> where T: PrivTr = T;
   |     ^^^^^^^^^^^^^^^^^ type alias `traits_where::Alias` is reachable at visibility `pub(crate)`
   |
note: but trait `traits_where::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:59:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits_where::PrivTr` is more private than the item `traits_where::Tr2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:66:5
   |
LL |     pub trait Tr2<T> where T: PrivTr {}
   |     ^^^^^^^^^^^^^^^^ trait `traits_where::Tr2` is reachable at visibility `pub(crate)`
   |
note: but trait `traits_where::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:59:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits_where::PrivTr` is more private than the item `traits_where::Tr3::f`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:69:9
   |
LL |         fn f<T>(arg: T) where T: PrivTr {}
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated function `traits_where::Tr3::f` is reachable at visibility `pub(crate)`
   |
note: but trait `traits_where::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:59:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `traits_where::PrivTr` is more private than the item `traits_where::Pub<T>`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:72:5
   |
LL |     impl<T> Pub<T> where T: PrivTr {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation `traits_where::Pub<T>` is reachable at visibility `pub(crate)`
   |
note: but trait `traits_where::PrivTr` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:59:5
   |
LL |     trait PrivTr {}
   |     ^^^^^^^^^^^^

error: trait `generics::PrivTr<generics::Pub>` is more private than the item `generics::Tr1`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:84:5
   |
LL |     pub trait Tr1: PrivTr<Pub> {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `generics::Tr1` is reachable at visibility `pub(crate)`
   |
note: but trait `generics::PrivTr<generics::Pub>` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:80:5
   |
LL |     trait PrivTr<T> {}
   |     ^^^^^^^^^^^^^^^

error: type `generics::Priv` is more private than the item `generics::Tr2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:86:5
   |
LL |     pub trait Tr2: PubTr<Priv> {} //~ ERROR type `generics::Priv` is more private than the item `generics::Tr2`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `generics::Tr2` is reachable at visibility `pub(crate)`
   |
note: but type `generics::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:78:5
   |
LL |     struct Priv<T = u8>(T);
   |     ^^^^^^^^^^^^^^^^^^^

error: type `generics::Priv` is more private than the item `generics::Tr3`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:87:5
   |
LL |     pub trait Tr3: PubTr<[Priv; 1]> {} //~ ERROR type `generics::Priv` is more private than the item `generics::Tr3`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `generics::Tr3` is reachable at visibility `pub(crate)`
   |
note: but type `generics::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:78:5
   |
LL |     struct Priv<T = u8>(T);
   |     ^^^^^^^^^^^^^^^^^^^

error: type `generics::Priv` is more private than the item `Tr4`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:88:5
   |
LL |     pub trait Tr4: PubTr<Pub<Priv>> {} //~ ERROR type `generics::Priv` is more private than the item `Tr4`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `Tr4` is reachable at visibility `pub(crate)`
   |
note: but type `generics::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:78:5
   |
LL |     struct Priv<T = u8>(T);
   |     ^^^^^^^^^^^^^^^^^^^

error[E0446]: private type `impls::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:119:9
   |
LL |     struct Priv;
   |     ----------- `impls::Priv` declared as private
...
LL |         type Alias = Priv; //~ ERROR private type `impls::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error: type `aliases_pub::Priv` is more private than the item `aliases_pub::<impl Pub2>::f`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:190:9
   |
LL |         pub fn f(arg: Priv) {} //~ ERROR type `aliases_pub::Priv` is more private than the item `aliases_pub::<impl Pub2>::f`
   |         ^^^^^^^^^^^^^^^^^^^ associated function `aliases_pub::<impl Pub2>::f` is reachable at visibility `pub(crate)`
   |
note: but type `aliases_pub::Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:163:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^

error[E0446]: private type `aliases_pub::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:193:9
   |
LL |     struct Priv;
   |     ----------- `aliases_pub::Priv` declared as private
...
LL |         type Check = Priv; //~ ERROR private type `aliases_pub::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error[E0446]: private type `aliases_pub::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:196:9
   |
LL |     struct Priv;
   |     ----------- `aliases_pub::Priv` declared as private
...
LL |         type Check = Priv; //~ ERROR private type `aliases_pub::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error[E0446]: private type `aliases_pub::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:199:9
   |
LL |     struct Priv;
   |     ----------- `aliases_pub::Priv` declared as private
...
LL |         type Check = Priv; //~ ERROR private type `aliases_pub::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error[E0446]: private type `aliases_pub::Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:202:9
   |
LL |     struct Priv;
   |     ----------- `aliases_pub::Priv` declared as private
...
LL |         type Check = Priv; //~ ERROR private type `aliases_pub::Priv` in public interface
   |         ^^^^^^^^^^ can't leak private type

error: trait `PrivTr1` is more private than the item `aliases_priv::Tr1`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:232:5
   |
LL |     pub trait Tr1: PrivUseAliasTr {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `aliases_priv::Tr1` is reachable at visibility `pub(crate)`
   |
note: but trait `PrivTr1` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:218:5
   |
LL |     trait PrivTr1<T = u8> {
   |     ^^^^^^^^^^^^^^^^^^^^^

error: trait `PrivTr1<Priv2>` is more private than the item `aliases_priv::Tr2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:234:5
   |
LL |     pub trait Tr2: PrivUseAliasTr<PrivAlias> {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `aliases_priv::Tr2` is reachable at visibility `pub(crate)`
   |
note: but trait `PrivTr1<Priv2>` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:218:5
   |
LL |     trait PrivTr1<T = u8> {
   |     ^^^^^^^^^^^^^^^^^^^^^

error: type `Priv2` is more private than the item `aliases_priv::Tr2`
##[error]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:234:5
   |
LL |     pub trait Tr2: PrivUseAliasTr<PrivAlias> {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trait `aliases_priv::Tr2` is reachable at visibility `pub(crate)`
   |
note: but type `Priv2` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:216:5
   |
LL |     struct Priv2;
   |     ^^^^^^^^^^^^

warning: bounds on generic parameters in type aliases are not enforced
##[warning]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:42:23
   |
LL |     pub type Alias<T: PrivTr> = T; //~ ERROR trait `traits::PrivTr` is more private than the item `traits::Alias`
   |                     --^^^^^^
   |                     | |
   |                     | will not be checked at usage sites of the type alias
   |                     help: remove this bound
   |
   = note: this is a known limitation of the type checker that may be lifted in a future edition.
           see issue #112792 <https://github.com/rust-lang/rust/issues/112792> for more information
   = help: add `#![feature(lazy_type_alias)]` to the crate attributes to enable the desired semantics
   = note: `#[warn(type_alias_bounds)]` on by default

warning: where clauses on type aliases are not enforced
##[warning]  --> /checkout/tests/ui/privacy/private-in-public-warn.rs:63:29
   |
LL |     pub type Alias<T> where T: PrivTr = T;
   |                       ------^^^^^^^^^
   |                       |     |
   |                       |     will not be checked at usage sites of the type alias
   |                       help: remove this where clause
   |
---

- warning: type `Priv` is more private than the item `Leak`
-   --> $DIR/projections.rs:3:5
-    |
- LL |     pub type Leak = Priv;
-    |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
-    |
- note: but type `Priv` is only usable at visibility `pub(self)`
-   --> $DIR/projections.rs:2:5
-    |
- LL |     struct Priv;
-    |     ^^^^^^^^^^^
-    = note: `#[warn(private_interfaces)]` on by default
- 
14 error[E0446]: private type `Priv` in public interface
15   --> $DIR/projections.rs:24:5
16    |

28 ...
29 LL |     type A<T: Trait> = T::A<m::Leak>;
30    |     ^^^^^^^^^^^^^^^^ can't leak private type
+ 
+ warning: type `Priv` is more private than the item `Leak`
+   --> $DIR/projections.rs:3:5
+    |
+ LL |     pub type Leak = Priv;
+    |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
+    |
+ note: but type `Priv` is only usable at visibility `pub(self)`
+   --> $DIR/projections.rs:2:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+    = note: `#[warn(private_interfaces)]` on by default
31 
32 error: type `Priv` is private
33   --> $DIR/projections.rs:14:15

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/privacy/projections.rs:3:5
- LL |     pub type Leak = Priv; //~ WARN: `Priv` is more private than the item `Leak`
-   --> /checkout/tests/ui/privacy/projections.rs:2:5
+ 
+ warning: type `Priv` is more private than the item `Leak`
+   --> $DIR/projections.rs:3:5
+    |
+ LL |     pub type Leak = Priv;
+    |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
+    |
+ note: but type `Priv` is only usable at visibility `pub(self)`
+   --> $DIR/projections.rs:2:5
+    |
+ LL |     struct Priv;
+    |     ^^^^^^^^^^^
+    = note: `#[warn(private_interfaces)]` on by default
---
To only update this specific test, also pass `--test-args privacy/projections.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/privacy/projections.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/projections" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0446]: private type `Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/projections.rs:24:5
   |
LL |     struct Priv;
   |     ----------- `Priv` declared as private
...
LL |     type A<T> = m::Leak;
   |     ^^^^^^^^^ can't leak private type

error[E0446]: private type `Priv` in public interface
##[error]  --> /checkout/tests/ui/privacy/projections.rs:39:5
   |
LL |     struct Priv;
   |     ----------- `Priv` declared as private
...
LL |     type A<T: Trait> = T::A<m::Leak>;
   |     ^^^^^^^^^^^^^^^^ can't leak private type

warning: type `Priv` is more private than the item `Leak`
##[warning]  --> /checkout/tests/ui/privacy/projections.rs:3:5
   |
LL |     pub type Leak = Priv; //~ WARN: `Priv` is more private than the item `Leak`
   |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
   |
note: but type `Priv` is only usable at visibility `pub(self)`
  --> /checkout/tests/ui/privacy/projections.rs:2:5
   |
LL |     struct Priv;
   |     ^^^^^^^^^^^
   = note: `#[warn(private_interfaces)]` on by default

error: type `Priv` is private
##[error]  --> /checkout/tests/ui/privacy/projections.rs:14:15
   |
LL | fn check() -> <u8 as Trait>::A<m::Leak> {
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^ private type

error: type `Priv` is private
##[error]  --> /checkout/tests/ui/privacy/projections.rs:29:39
   |
LL |   fn check2() -> <u8 as Trait2>::A<u32> {
   |  _______________________________________^
LL | |     //~^ ERROR: `Priv` is private
LL | |     todo!()
LL | | }
   | |_^ private type

error: type `Priv` is private
##[error]  --> /checkout/tests/ui/privacy/projections.rs:24:17
   |
LL |     type A<T> = m::Leak;
   |                 ^^^^^^^ private type

error: type `Priv` is private
##[error]  --> /checkout/tests/ui/privacy/projections.rs:39:24
   |
LL |     type A<T: Trait> = T::A<m::Leak>;
   |                        ^^^^^^^^^^^^^ private type

error: aborting due to 6 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0446`.
```

</details>

--- a/compiler/rustc_interface/src/passes.rs
+++ b/compiler/rustc_interface/src/passes.rs
@@ -1147,7 +1147,9 @@ fn analysis(tcx: TyCtxt<'_>, (): ()) {
 
                 parallel!(
                     {
-                        tcx.ensure_ok().check_private_in_public(());
+                        tcx.par_hir_for_each_module(|module| {
+                            tcx.ensure_ok().check_private_in_public(module)
+                        })
                     },
                     {
                         tcx.par_hir_for_each_module(|module| {

--- a/compiler/rustc_middle/src/query/mod.rs
+++ b/compiler/rustc_middle/src/query/mod.rs
@@ -1390,8 +1390,11 @@ rustc_queries! {
         eval_always
         desc { "checking effective visibilities" }
     }
-    query check_private_in_public(_: ()) {
-        desc { "checking for private elements in public interfaces" }
+    query check_private_in_public(module_def_id: LocalModDefId) {
+        desc { |tcx|
+            "checking for private elements in public interfaces for {}",
+            describe_as_module(module_def_id, tcx)
+        }
     }
 
     query reachable_set(_: ()) -> &'tcx LocalDefIdSet {

--- a/compiler/rustc_privacy/src/lib.rs
+++ b/compiler/rustc_privacy/src/lib.rs
@@ -1854,12 +1854,12 @@ fn effective_visibilities(tcx: TyCtxt<'_>, (): ()) -> &EffectiveVisibilities {
     tcx.arena.alloc(visitor.effective_visibilities)
 }
 
-fn check_private_in_public(tcx: TyCtxt<'_>, (): ()) {
+fn check_private_in_public(tcx: TyCtxt<'_>, module_def_id: LocalModDefId) {
     let effective_visibilities = tcx.effective_visibilities(());
     // Check for private types in public interfaces.
     let mut checker = PrivateItemsInPublicInterfacesChecker { tcx, effective_visibilities };
 
-    let crate_items = tcx.hir_crate_items(());
+    let crate_items = tcx.hir_module_items(module_def_id);
     for id in crate_items.free_items() {
         checker.check_item(id);
     }

--- a/compiler/rustc_privacy/src/lib.rs
+++ b/compiler/rustc_privacy/src/lib.rs
@@ -1423,8 +1423,6 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
         };
 
         let vis = self.tcx.local_visibility(local_def_id);
-        let span = self.tcx.def_span(self.item_def_id.to_def_id());
-        let vis_span = self.tcx.def_span(def_id);
         if self.in_assoc_ty && !vis.is_at_least(self.required_visibility, self.tcx) {
             let vis_descr = match vis {
                 ty::Visibility::Public => "public",
@@ -1441,6 +1439,8 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
                 }
             };
 
+            let span = self.tcx.def_span(self.item_def_id.to_def_id());
+            let vis_span = self.tcx.def_span(def_id);
             self.tcx.dcx().emit_err(InPublicInterface {
                 span,
                 vis_descr,
@@ -1463,6 +1463,8 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
             } else {
                 lint::builtin::PRIVATE_BOUNDS
             };
+            let span = self.tcx.def_span(self.item_def_id.to_def_id());
+            let vis_span = self.tcx.def_span(def_id);
             self.tcx.emit_node_span_lint(
                 lint,
                 self.tcx.local_def_id_to_hir_id(self.item_def_id),

--- a/compiler/rustc_privacy/src/lib.rs
+++ b/compiler/rustc_privacy/src/lib.rs
@@ -1594,7 +1594,7 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'_, 'tcx> {
         self.effective_visibilities.effective_vis(def_id).copied()
     }
 
-    fn check_item(&mut self, id: ItemId) {
+    fn check_item(&self, id: ItemId) {
         let tcx = self.tcx;
         let def_id = id.owner_id.def_id;
         let item_visibility = tcx.local_visibility(def_id);
@@ -1722,7 +1722,7 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'_, 'tcx> {
         }
     }
 
-    fn check_foreign_item(&mut self, id: ForeignItemId) {
+    fn check_foreign_item(&self, id: ForeignItemId) {
         let tcx = self.tcx;
         let def_id = id.owner_id.def_id;
         let item_visibility = tcx.local_visibility(def_id);
@@ -1857,13 +1857,9 @@ fn effective_visibilities(tcx: TyCtxt<'_>, (): ()) -> &EffectiveVisibilities {
 fn check_private_in_public(tcx: TyCtxt<'_>, module_def_id: LocalModDefId) {
     let effective_visibilities = tcx.effective_visibilities(());
     // Check for private types in public interfaces.
-    let mut checker = PrivateItemsInPublicInterfacesChecker { tcx, effective_visibilities };
+    let checker = PrivateItemsInPublicInterfacesChecker { tcx, effective_visibilities };
 
     let crate_items = tcx.hir_module_items(module_def_id);
-    for id in crate_items.free_items() {
-        checker.check_item(id);
-    }
-    for id in crate_items.foreign_items() {
-        checker.check_foreign_item(id);
-    }
+    let _ = crate_items.par_items(|id| Ok(checker.check_item(id)));
+    let _ = crate_items.par_foreign_items(|id| Ok(checker.check_foreign_item(id)));
 }

--- a/tests/ui/privacy/private-in-public-warn.stderr
+++ b/tests/ui/privacy/private-in-public-warn.stderr
@@ -93,6 +93,42 @@ LL |     struct Priv;
 LL |         type Alias = Priv;
    |         ^^^^^^^^^^ can't leak private type
 
+error: type `types::Priv` is more private than the item `types::ES`
+  --> $DIR/private-in-public-warn.rs:27:9
+   |
+LL |         pub static ES: Priv;
+   |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
+   |
+note: but type `types::Priv` is only usable at visibility `pub(self)`
+  --> $DIR/private-in-public-warn.rs:9:5
+   |
+LL |     struct Priv;
+   |     ^^^^^^^^^^^
+
+error: type `types::Priv` is more private than the item `types::ef1`
+  --> $DIR/private-in-public-warn.rs:28:9
+   |
+LL |         pub fn ef1(arg: Priv);
+   |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
+   |
+note: but type `types::Priv` is only usable at visibility `pub(self)`
+  --> $DIR/private-in-public-warn.rs:9:5
+   |
+LL |     struct Priv;
+   |     ^^^^^^^^^^^
+
+error: type `types::Priv` is more private than the item `types::ef2`
+  --> $DIR/private-in-public-warn.rs:29:9
+   |
+LL |         pub fn ef2() -> Priv;
+   |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
+   |
+note: but type `types::Priv` is only usable at visibility `pub(self)`
+  --> $DIR/private-in-public-warn.rs:9:5
+   |
+LL |     struct Priv;
+   |     ^^^^^^^^^^^
+
 error: trait `traits::PrivTr` is more private than the item `traits::Alias`
   --> $DIR/private-in-public-warn.rs:42:5
    |
@@ -358,42 +394,6 @@ note: but type `Priv2` is only usable at visibility `pub(self)`
    |
 LL |     struct Priv2;
    |     ^^^^^^^^^^^^
-
-error: type `types::Priv` is more private than the item `types::ES`
-  --> $DIR/private-in-public-warn.rs:27:9
-   |
-LL |         pub static ES: Priv;
-   |         ^^^^^^^^^^^^^^^^^^^ static `types::ES` is reachable at visibility `pub(crate)`
-   |
-note: but type `types::Priv` is only usable at visibility `pub(self)`
-  --> $DIR/private-in-public-warn.rs:9:5
-   |
-LL |     struct Priv;
-   |     ^^^^^^^^^^^
-
-error: type `types::Priv` is more private than the item `types::ef1`
-  --> $DIR/private-in-public-warn.rs:28:9
-   |
-LL |         pub fn ef1(arg: Priv);
-   |         ^^^^^^^^^^^^^^^^^^^^^^ function `types::ef1` is reachable at visibility `pub(crate)`
-   |
-note: but type `types::Priv` is only usable at visibility `pub(self)`
-  --> $DIR/private-in-public-warn.rs:9:5
-   |
-LL |     struct Priv;
-   |     ^^^^^^^^^^^
-
-error: type `types::Priv` is more private than the item `types::ef2`
-  --> $DIR/private-in-public-warn.rs:29:9
-   |
-LL |         pub fn ef2() -> Priv;
-   |         ^^^^^^^^^^^^^^^^^^^^^ function `types::ef2` is reachable at visibility `pub(crate)`
-   |
-note: but type `types::Priv` is only usable at visibility `pub(self)`
-  --> $DIR/private-in-public-warn.rs:9:5
-   |
-LL |     struct Priv;
-   |     ^^^^^^^^^^^
 
 warning: bounds on generic parameters in type aliases are not enforced
   --> $DIR/private-in-public-warn.rs:42:23

--- a/tests/ui/privacy/projections.stderr
+++ b/tests/ui/privacy/projections.stderr
@@ -1,16 +1,3 @@
-warning: type `Priv` is more private than the item `Leak`
-  --> $DIR/projections.rs:3:5
-   |
-LL |     pub type Leak = Priv;
-   |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
-   |
-note: but type `Priv` is only usable at visibility `pub(self)`
-  --> $DIR/projections.rs:2:5
-   |
-LL |     struct Priv;
-   |     ^^^^^^^^^^^
-   = note: `#[warn(private_interfaces)]` on by default
-
 error[E0446]: private type `Priv` in public interface
   --> $DIR/projections.rs:24:5
    |
@@ -28,6 +15,19 @@ LL |     struct Priv;
 ...
 LL |     type A<T: Trait> = T::A<m::Leak>;
    |     ^^^^^^^^^^^^^^^^ can't leak private type
+
+warning: type `Priv` is more private than the item `Leak`
+  --> $DIR/projections.rs:3:5
+   |
+LL |     pub type Leak = Priv;
+   |     ^^^^^^^^^^^^^ type alias `Leak` is reachable at visibility `pub(crate)`
+   |
+note: but type `Priv` is only usable at visibility `pub(self)`
+  --> $DIR/projections.rs:2:5
+   |
+LL |     struct Priv;
+   |     ^^^^^^^^^^^
+   = note: `#[warn(private_interfaces)]` on by default
 
 error: type `Priv` is private
   --> $DIR/projections.rs:14:15

